<a href="https://colab.research.google.com/github/EhsanHonarvar/Practices/blob/main/1_StepByStep.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **from:**
https://github.com/theCaiGuy/Twitter-Sentiment


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd

In [ ]:
# ########## With Bitcoin Tweets Data ################ Start
# tweets = pd.read_csv('/kaggle/input/bitcoin-tweets-16m-tweets-with-sentiment-tagged/mbsa.csv')
# tweets.head()
# ########## With Bitcoin Tweets Data ################ End


full_data = pd.read_csv("/content/drive/MyDrive/Thesis/training.1600000.processed.noemoticon.csv", names= ['Pos_Neg','id','date','query_string','user','Content'] , encoding = 'latin-1')
full_n = full_data.shape[0]
full_data.head()

In [ ]:
print(type(full_data))
full_data.head(5)

<class 'pandas.core.frame.DataFrame'>


Pos_Neg          id                          date query_string  \
0        0  1467810369  Mon Apr 06 22:19:45 PDT 2009     NO_QUERY   
1        0  1467810672  Mon Apr 06 22:19:49 PDT 2009     NO_QUERY   
2        0  1467810917  Mon Apr 06 22:19:53 PDT 2009     NO_QUERY   
3        0  1467811184  Mon Apr 06 22:19:57 PDT 2009     NO_QUERY   
4        0  1467811193  Mon Apr 06 22:19:57 PDT 2009     NO_QUERY   

              user                                            Content  
0  _TheSpecialOne_  @switchfoot http://twitpic.com/2y1zl - Awww, t...  
1    scotthamilton  is upset that he can't update his Facebook by ...  
2         mattycus  @Kenichan I dived many times for the ball. Man...  
3          ElleCTF    my whole body feels itchy and like its on fire   
4           Karoli  @nationwideclass no, it's not behaving at all....

In [ ]:
full_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1600000 entries, 0 to 1599999
Data columns (total 6 columns):
 #   Column        Non-Null Count    Dtype 
---  ------        --------------    ----- 
 0   Pos_Neg       1600000 non-null  int64 
 1   id            1600000 non-null  int64 
 2   date          1600000 non-null  object
 3   query_string  1600000 non-null  object
 4   user          1600000 non-null  object
 5   Content       1600000 non-null  object
dtypes: int64(2), object(4)
memory usage: 73.2+ MB


In [ ]:
import sklearn
from sklearn.model_selection import train_test_split

In [ ]:
train_data, dev_data = sklearn.model_selection.train_test_split(full_data, test_size = 0.05)

In [ ]:
print(type(train_data))

<class 'pandas.core.frame.DataFrame'>


In [ ]:
print(full_n)

1600000


In [ ]:
# full_data = full_data [:320000]

In [ ]:
# full_n = full_data.shape[0]
# print(full_n)

320000


In [ ]:
# train_data, dev_data = sklearn.model_selection.train_test_split(full_data, test_size = 0.40) #It was 0.04

In [ ]:
# train_data = train_data[:96000]
# dev_data = dev_data[:64000]

In [ ]:
print(full_data.shape[0],'\n',train_data.shape[0],'\n',dev_data.shape[0])

1600000 
 1520000 
 80000


In [ ]:

# Get ground truth x and y values
train_x_raw = train_data.loc[:]["Content"]
train_y = [0.0 if y == 0 else 1.0 for y in train_data.loc[:]["Pos_Neg"]]

dev_x_raw = dev_data.loc[:]["Content"]
dev_y = [0.0 if y == 0 else 1.0 for y in dev_data.loc[:]["Pos_Neg"]]
#print (dev_y)

print(type(train_x_raw))
print(type(dev_y))

<class 'pandas.core.series.Series'>
<class 'list'>


In [ ]:
# import gensim.downloader

In [ ]:
# glove_vectors = gensim.downloader.load('glove-twitter-100')

[==================================================] 100.0% 387.1/387.1MB downloaded


In [ ]:
# print(glove_vectors)

In [ ]:
char_vectors = {}
for line in open('/content/drive/MyDrive/Datasets/glove.6B.100d.txt').readlines():
    sp = line.strip().split()
    if len(sp) == 0: continue
    char_vectors[sp[0]] = [float(x) for x in sp[1:]]

print(type(char_vectors))
print(type(char_vectors[sp[0]]))

<class 'dict'>
<class 'list'>


In [ ]:
print(len(char_vectors))
print(len(char_vectors[sp[0]]))

400000
100


In [ ]:
# print(char_vectors)

print(char_vectors[sp[0]])
print(char_vectors['i'])
print(char_vectors['the'])

[0.28365, -0.6263, -0.44351, 0.2177, -0.087421, -0.17062, 0.29266, -0.024899, 0.26414, -0.17023, 0.25817, 0.097484, -0.33103, -0.43859, 0.0095799, 0.095624, -0.17777, 0.38886, 0.27151, 0.14742, -0.43973, -0.26588, -0.024271, 0.27186, -0.36761, -0.24827, -0.20815, 0.22128, -0.044409, 0.021373, 0.24594, 0.26143, 0.29303, 0.13281, 0.082232, -0.12869, 0.1622, -0.22567, -0.060348, 0.28703, 0.11381, 0.34839, 0.3419, 0.36996, -0.13592, 0.0062694, 0.080317, 0.0036251, 0.43093, 0.01882, 0.31008, 0.16722, 0.074112, -0.37745, 0.47363, 0.41284, 0.24471, 0.075965, -0.51725, -0.49481, 0.526, -0.074645, 0.41434, -0.1956, -0.16544, -0.045649, -0.40153, -0.13136, -0.4672, 0.18825, 0.2612, 0.16854, 0.22615, 0.62992, -0.1288, 0.055841, 0.01928, 0.024572, 0.46875, 0.2582, -0.31672, 0.048591, 0.3277, -0.50141, 0.30855, 0.11997, -0.25768, -0.039867, -0.059672, 0.5525, 0.13885, -0.22862, 0.071792, -0.43208, 0.5398, -0.085806, 0.032651, 0.43678, -0.82607, -0.15701]
[-0.046539, 0.61966, 0.56647, -0.46584, -1.1

In [ ]:
print(char_vectors['iran'])

[0.73704, 0.66283, 1.0574, 0.010218, 0.39815, -1.312, -0.82644, 0.52181, 0.71802, 1.4257, 0.32899, -0.70814, 1.0913, 0.39288, -0.539, 0.34568, -0.44541, -0.68467, -1.0251, -0.28259, 0.24643, -0.12216, 0.6778, 0.037953, -0.49667, 0.22861, -0.42924, 0.82565, 0.92749, 0.82861, -0.29846, 0.078323, -0.6415, -0.049715, 0.097872, 0.11647, 0.50231, -0.66332, -0.8046, -0.13047, -1.3995, 0.16397, 0.51603, -0.53419, 0.20498, -1.3766, 0.94354, 0.061697, -0.050002, -0.48968, 0.7309, 0.99181, 0.44085, 1.4769, -0.9737, -1.8742, 0.13894, -0.84341, 0.75602, -0.35825, -0.079888, -0.50491, 0.083148, -0.71243, 0.65236, 0.6766, 0.28151, 0.75826, -0.65151, 0.55292, 0.84885, -0.081808, 0.030953, -0.18909, 0.50983, 0.19043, 0.25679, -0.16961, -0.8492, 0.30048, 1.5914, 0.16726, -0.28329, -0.49085, -0.93667, 0.33378, 0.0065908, 0.76209, 0.44116, -0.77764, -0.44926, 0.57403, -0.44588, -0.31071, -0.91459, 0.31412, -0.343, -0.27904, 0.68867, -0.17311]


In [ ]:
import string
import nltk
from nltk.tokenize import word_tokenize
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [ ]:

SEED = 1234
UNK = '<unk>'
PAD = '<pad>'

TWEET_LEN = 20
EMBED_LEN = 100
N_EPOCHS = 70 #4000


In [ ]:
#>Creation of tok2id dictionary base on Contents
tok2id = {}
for ex in full_data['Content']:
    for w in word_tokenize(ex):
        if w in string.punctuation:
            continue
        if not w in tok2id:
            tok2id[w] = len(tok2id)

tok2id[UNK] = len(tok2id)
tok2id[PAD] = len(tok2id)

In [ ]:
len(tok2id)

875221

In [ ]:
#>Function for Tokenization of Sentences as Vector of tokens
def vectorize(examples, tok2id):
    vec_examples = []
    for ex in examples:
        #print (ex)
        sentence = []
        for w in word_tokenize(ex):
            if w in string.punctuation:
                continue
            if w in tok2id:
                sentence.append(tok2id[w])
        if len(sentence) < TWEET_LEN:
            sentence += [tok2id[PAD] for i in range(TWEET_LEN - len(sentence))]
        else:
            sentence = sentence[:TWEET_LEN]
        vec_examples.append(sentence)
    return vec_examples

In [ ]:
train_x = vectorize(train_x_raw, tok2id)

dev_x = vectorize(dev_x_raw, tok2id)



In [ ]:
from google.colab import files

In [ ]:
# files.upload()

Saving TweetDataset.py to TweetDataset.py


{'TweetDataset.py': b'from torch.utils.data import Dataset\n\nclass TweetDataset(Dataset):\n    """Face Landmarks dataset."""\n\n    def __init__(self, X, Y):\n        self.x = X\n        self.y = Y\n\n    def __len__(self):\n        return len(self.x)\n\n    def __getitem__(self, idx):\n        content = self.x[idx]\n        label = self.y[idx]\n        sample = {\'content\': content, \'label\': label}\n\n        return sample\n\ndataset = TweetDataset(None, None)\n'}

In [ ]:
# import TweetDataset
# import numpy as np

In [ ]:
###Instead Above##
# from torch.utils.data import Dataset

# class TweetDataset(Dataset):
#     """Face Landmarks dataset."""

#     def __init__(self, X, Y):
#         self.x = X
#         self.y = Y

#     def __len__(self):
#         return len(self.x)

#     def __getitem__(self, idx):
#         content = self.x[idx]
#         label = self.y[idx]
#         sample = {'content': content, 'label': label}

#         return sample

# dataset = TweetDataset(None, None)

# **#?#**

In [ ]:

from torch.utils.data import Dataset

class TweetDataset(Dataset):
    """Face Landmarks dataset."""

    def __init__(self, X, Y):
        self.x = X
        self.y = Y

    def __len__(self):
        return len(self.x)

    def __getitem__(self, idx):
        content = self.x[idx]
        label = self.y[idx]
        sample = {'content': content, 'label': label}

        return sample

dataset = TweetDataset(None, None)

In [ ]:
#>Definition of Function to Creat embedding_matrix based on tokens and GLOVE embedding vectors
import numpy as np
def build_embeddings(tok2id, char_vectors):
    embeddings_matrix = np.asarray(np.random.normal(0, 0.9, (len(tok2id), EMBED_LEN)), dtype='float32')

    for token in tok2id:
        i = tok2id[token]
        if token in char_vectors:
            embeddings_matrix[i] = char_vectors[token]
        elif token.lower() in char_vectors:
            embeddings_matrix[i] = char_vectors[token.lower()]

    return embeddings_matrix

In [ ]:
train_dataset = TweetDataset(train_x, train_y)
dev_dataset = TweetDataset(dev_x, dev_y)

embeddings = build_embeddings(tok2id, char_vectors)

In [ ]:
print(len(embeddings))
print(len(embeddings[0]))

875221
100


In [ ]:
# files.upload() # convblock, convblockdeeper, DeepCNN, DeeperCnn

# **Building Conv Blocks**

In [ ]:
import sys
from typing import List, Tuple, Dict, Set, Union
import torch
import torch.nn as nn
import torch.nn.utils
import torch.nn.functional as F
import torch.optim as optim
import numpy as np


In [ ]:
class ConvBlock(nn.Module):
    def __init__(self, in_channels, num_filters, k = 5, stride = 1):
        """ Initialize a CNN network with a kernel of size k,  """
        super().__init__()

        self.in_channels = in_channels
        self.num_filters = num_filters
        self.k = k
        self.stride = stride

        self.conv1 = nn.Conv1d(in_channels = in_channels, out_channels = num_filters, kernel_size = k, bias = True, padding = 2)
        self.batch_norm1 = nn.BatchNorm1d(num_filters)
        self.relu1 = nn.ReLU()
        #torch.nn.init.xavier_uniform_(self.conv1.weight)

        self.conv2 = nn.Conv1d(in_channels = num_filters, out_channels = num_filters, kernel_size = k, bias = True, padding = 2)
        self.batch_norm2 = nn.BatchNorm1d(num_filters)
        self.relu2 = nn.ReLU()
        #torch.nn.init.xavier_uniform_(self.conv2.weight)

        self.pool = nn.MaxPool1d(kernel_size = 2)

    def forward(self, x, print_sizes = False) -> torch.Tensor:
        """ Input x of size (batch size, embed_size, sent_len) and output of size [batchsize, embed_size] """

        x_conv1 = self.conv1(x)
        x_relu1 = self.relu1(x_conv1)
        x_bn = self.batch_norm1(x_relu1)
        if print_sizes: print (x_bn.shape)

        x_conv2 = self.batch_norm2(self.relu2(self.conv2(x_bn)))
        if print_sizes: print (x_conv2.shape)

        x_pool = self.pool(x_conv2)
        # print (x_pool)
        if print_sizes: print(x_pool.shape)

        return x_pool

# def test_all():
#     batch_size = 3
#     embed_size = 5
#     sent_len = 6
#     num_filters = 6
#
#     fake_input = torch.tensor(np.zeros((batch_size, embed_size, sent_len)), dtype = torch.float32)
#     block = ConvBlock(embed_size, num_filters)
#     print (block.forward(fake_input, True))
#
#
# if __name__ == '__main__':
#     #test_intermediate_sizes()
#     #test_output()
#     test_all()

In [ ]:
class DeepCNN(nn.Module):
    def __init__(self, embeddings, dropout=0.5):
        super().__init__()

        self.batch_size = embeddings.shape[1]

        # Freeze pretrained GloVe embeddings
        self.embedding = nn.Embedding(embeddings.shape[0], embeddings.shape[1])
        self.embedding.weight = nn.Parameter(torch.tensor(embeddings))
        self.embedding.weight.requires_grad = False

        self.conv = nn.Conv1d(in_channels=embeddings.shape[1], out_channels=128, kernel_size=5, padding = 2)
        #torch.nn.init.xavier_uniform_(self.conv.weight)
        self.block = ConvBlock(128, 256)
        self.block2 = ConvBlock(256, 256)

        self.dropout = nn.Dropout(dropout)
        self.fc = nn.Linear(5 * 256, 256)
        self.dropout2 = nn.Dropout(dropout)
        self.fc2 = nn.Linear(256, 1)
        #torch.nn.init.xavier_uniform_(self.fc.weight)

    def forward(self, x, print_sizes = False):
        # x.shape = (sent_len, batch_size)
        x = x.permute(1, 0)
        if print_sizes: print ("x.shape: " + str(x.shape))
        # x.shape = (batch_size, sent_len)

        x_emb = self.embedding(x).permute(0, 2, 1)
        if print_sizes: print ("x_emb.shape: " + str(x_emb.shape))
        # x_emb.shape = (batch_size, emb_size, sent_len)

        x_conv = self.conv(x_emb)
        # x_conv.shape = (batch_size, 128, sent_len)
        if print_sizes: print ("x_conv.shape: " + str(x_conv.shape))

        x_block = self.block.forward(x_conv)
        # x_block.shape = (batch_size, 128, sent_len / 2)
        if print_sizes: print ("x_block.shape: " + str(x_block.shape))

        x_block_2 = self.block2.forward(x_block)
        # x_block.shape = (batch_size, 256, sent_len / 4)
        if print_sizes: print ("x_block_2.shape: " + str(x_block_2.shape))

        x_cat = x_block_2.view(-1, x_block_2.shape[1] * x_block_2.shape[2])
        # x_cat.shape = (batch_size, 128 * sent_len / 2)
        if print_sizes: print ("x_cat.shape: " + str(x_cat.shape))

        x_fc = self.fc(self.dropout(x_cat))
        # x_fc.shape = (batch_size, 1)
        if print_sizes: print ("x_fc.shape: " + str(x_fc.shape))

        x_fc2 = self.fc2(self.dropout2(x_fc))

        return x_fc2

In [ ]:
### convblockdeeper ###
# class ConvBlock(nn.Module):
#     def __init__(self, in_channels, num_filters, k = 5, stride = 1):
#         """ Initialize a CNN network with a kernel of size k,  """
#         super().__init__()

#         self.in_channels = in_channels
#         self.num_filters = num_filters
#         self.k = k
#         self.stride = stride

#         self.conv1 = nn.Conv1d(in_channels = in_channels, out_channels = num_filters, kernel_size = k, bias = True, padding = 2)
#         self.batch_norm1 = nn.BatchNorm1d(num_filters)
#         self.relu1 = nn.ReLU()

#         self.conv2 = nn.Conv1d(in_channels = num_filters, out_channels = num_filters, kernel_size = k, bias = True, padding = 2)
#         self.batch_norm2 = nn.BatchNorm1d(num_filters)
#         self.relu2 = nn.ReLU()

#         self.conv3 = nn.Conv1d(in_channels = num_filters, out_channels = num_filters, kernel_size = k, bias = True, padding = 2)
#         self.batch_norm3 = nn.BatchNorm1d(num_filters)
#         self.relu3 = nn.ReLU()

#         self.conv4 = nn.Conv1d(in_channels = num_filters, out_channels = num_filters, kernel_size = k, bias = True, padding = 2)
#         self.batch_norm4 = nn.BatchNorm1d(num_filters)
#         self.relu4 = nn.ReLU()

#         self.pool = nn.MaxPool1d(kernel_size = 2)

#     def forward(self, x, print_sizes = False) -> torch.Tensor:
#         """ Input x of size (batch size, embed_size, sent_len) and output of size [batchsize, embed_size] """

#         x_conv1 = self.conv1(x)
#         x_relu1 = self.relu1(x_conv1)
#         x_bn = self.batch_norm1(x_relu1)
#         if print_sizes: print (x_bn.shape)

#         x_conv2 = self.batch_norm2(self.relu2(self.conv2(x_bn)))
#         if print_sizes: print (x_conv2.shape)

#         x_conv3 = self.batch_norm3(self.relu3(self.conv3(x_conv2)))
#         if print_sizes: print (x_conv3.shape)

#         x_conv4 = self.batch_norm4(self.relu4(self.conv4(x_conv3)))
#         if print_sizes: print (x_conv4.shape)

#         x_pool = self.pool(x_conv4)
#         # print (x_pool)
#         if print_sizes: print(x_pool.shape)

#         return x_pool

# def test_all():
#     batch_size = 3
#     embed_size = 5
#     sent_len = 6
#     num_filters = 6

#     fake_input = torch.tensor(np.zeros((batch_size, embed_size, sent_len)), dtype = torch.float32)
#     block = ConvBlock(embed_size, num_filters)
#     print (block.forward(fake_input, True))
# #
# #
# if __name__ == '__main__':
# #     #test_intermediate_sizes()
# #     #test_output()
#     test_all()

torch.Size([3, 6, 6])
torch.Size([3, 6, 6])
torch.Size([3, 6, 6])
torch.Size([3, 6, 6])
torch.Size([3, 6, 3])
tensor([[[-0.2099,  0.5904,  0.4115],
         [ 0.0000,  0.0000,  0.0000],
         [-0.5126,  0.4798,  0.4700],
         [ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000],
         [-0.4663,  0.8142,  0.3347]],

        [[-0.2099,  0.5904,  0.4115],
         [ 0.0000,  0.0000,  0.0000],
         [-0.5126,  0.4798,  0.4700],
         [ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000],
         [-0.4663,  0.8142,  0.3347]],

        [[-0.2099,  0.5904,  0.4115],
         [ 0.0000,  0.0000,  0.0000],
         [-0.5126,  0.4798,  0.4700],
         [ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000],
         [-0.4663,  0.8142,  0.3347]]], grad_fn=<SqueezeBackward1>)


In [ ]:
### deeperCNN ###
# class DeepCNN(nn.Module):
#     def __init__(self, embeddings, dropout=0.5):
#         super().__init__()

#         self.batch_size = embeddings.shape[1]

#         # Freeze pretrained GloVe embeddings
#         self.embedding = nn.Embedding(embeddings.shape[0], embeddings.shape[1])
#         self.embedding.weight = nn.Parameter(torch.tensor(embeddings))
#         self.embedding.weight.requires_grad = False

#         self.conv = nn.Conv1d(in_channels=embeddings.shape[1], out_channels=128, kernel_size=5, padding = 2)
#         #torch.nn.init.xavier_uniform_(self.conv.weight)
#         self.block = ConvBlock(128, 256)
#         self.block2 = ConvBlock(256, 256)

#         self.dropout = nn.Dropout(dropout)
#         self.fc = nn.Linear(5 * 256, 256)
#         #self.relu = nn.ReLU()
#         self.dropout2 = nn.Dropout(dropout)
#         self.fc2 = nn.Linear(256, 1)
#         #torch.nn.init.xavier_uniform_(self.fc.weight)

#     def forward(self, x, print_sizes = False):
#         # x.shape = (sent_len, batch_size)
#         x = x.permute(1, 0)
#         if print_sizes: print ("x.shape: " + str(x.shape))
#         # x.shape = (batch_size, sent_len)

#         x_emb = self.embedding(x).permute(0, 2, 1)
#         if print_sizes: print ("x_emb.shape: " + str(x_emb.shape))
#         # x_emb.shape = (batch_size, emb_size, sent_len)

#         x_conv = self.conv(x_emb)
#         # x_conv.shape = (batch_size, 128, sent_len)
#         if print_sizes: print ("x_conv.shape: " + str(x_conv.shape))

#         x_block = self.block.forward(x_conv)
#         # x_block.shape = (batch_size, 128, sent_len / 2)
#         if print_sizes: print ("x_block.shape: " + str(x_block.shape))

#         x_block_2 = self.block2.forward(x_block)
#         # x_block.shape = (batch_size, 256, sent_len / 4)
#         if print_sizes: print ("x_block_2.shape: " + str(x_block_2.shape))

#         x_cat = x_block_2.view(-1, x_block_2.shape[1] * x_block_2.shape[2])
#         # x_cat.shape = (batch_size, 128 * sent_len / 2)
#         if print_sizes: print ("x_cat.shape: " + str(x_cat.shape))

#         x_fc = self.fc(self.dropout(x_cat))
#         # x_fc.shape = (batch_size, 1)
#         if print_sizes: print ("x_fc.shape: " + str(x_fc.shape))

#         x_fc2 = self.fc2(self.dropout2(x_fc))

#         return x_fc2

In [ ]:
# from DeepCNN import DeepCNN

In [ ]:
embeddings_matrix = embeddings #renaming

In [ ]:
import torch
import torch.nn as nn
import torch.nn.functional as F

from torch.utils.data import Dataset, DataLoader
from torchvision import transforms, utils

import sys
from typing import List, Tuple, Dict, Set, Union


import torch.nn.utils
import torch.optim as optim


import random
import os
import pandas as pd
import numpy as np

import string

#Ehsan 3 next lines
# from TweetDataset import TweetDataset
# from vocab import VocabEntry
# from convblock import ConvBlock

#from gensim.models import word2vec

SEED = 1234

In [ ]:
CNN_model = DeepCNN(embeddings_matrix)

# **#################################################**

In [ ]:
# device = torch.device('cpu') #('cuda' if torch.cuda.is_available() else 'cpu')

optimizer = optim.Adam(CNN_model.parameters())

criterion = nn.BCEWithLogitsLoss()

# CNN_model = CNN_model.to(device)

# criterion = criterion.to(device)

train_loader = DataLoader(train_dataset,
                      batch_size=128,
                      shuffle=True,
                      num_workers=4
                      # pin_memory=True # CUDA only
                      )

dev_loader = DataLoader(dev_dataset,
                  batch_size=128,
                  shuffle=False,
                  num_workers=4
                  # pin_memory=True # CUDA only
                  )

best_valid_acc = 0.0

In [ ]:
def binary_accuracy(preds, y):
    """
    Returns accuracy per batch, i.e. if you get 8/10 right, this returns 0.8, NOT 8
    """

    #round predictions to the closest integer
    rounded_preds = torch.round(torch.sigmoid(preds))
    correct = (rounded_preds == y).float() #convert into float for division
    acc = correct.sum()/len(correct)
    return acc

In [ ]:
def evaluate(model, dev_loader, criterion):

    epoch_loss = 0
    epoch_acc = 0

    model.eval()

    with torch.no_grad():

        for batchnum, batch in enumerate(dev_loader):
            dev_x = torch.stack(batch['content'])
            #print (train_x)
            dev_y = batch['label'].float()

            predictions = model(dev_x).squeeze(1)
            #print (torch.round(predictions))

            loss = criterion(predictions, dev_y)

            acc = binary_accuracy(predictions, dev_y)

            epoch_loss += loss.item()
            epoch_acc += acc.item()

    return epoch_loss / len(dev_loader), epoch_acc / len(dev_loader)

In [ ]:
def train(model, train_loader, optimizer, criterion):

    epoch_loss = 0
    epoch_acc = 0

    model.train()

    for batchnum, batch in enumerate(train_loader):
        #print ("Training on batch #" + str(batchnum))
        train_x = torch.stack(batch['content'])
        #print (train_x.shape)
        train_y = batch['label'].float()
        #train_y = batch['label'].long()
        if train_x.shape[1] == 1: continue
        #print (train_y.view(-1).shape)

        predictions = model.forward(train_x).squeeze(1)
        #print (predictions.shape)
        loss = criterion(predictions, train_y)
        # print (loss)
        loss.backward()
        grad_norm = torch.nn.utils.clip_grad_norm_(model.parameters(), 5)
        optimizer.step()
        acc = binary_accuracy(predictions, train_y)
        optimizer.step()

        epoch_loss += loss.item()
        epoch_acc += acc.item()

    return epoch_loss / len(train_loader), epoch_acc / len(train_loader)

In [ ]:
# MODEL_PATH = '/content/drive/MyDrive/models/cache'
# MODEL_PATH = '/content/drive/MyDrive/models/model_cache/cache'
MODEL_PATH = './cache'
# MODEL_PATH = './model_cache/cache'
N_EPOCHS = 70 # 20 #It was 4000

In [ ]:
best_valid_acc = 0

train_loss_values = np.zeros(N_EPOCHS)
train_acc_values = np.zeros(N_EPOCHS)

valid_loss_values = np.zeros(N_EPOCHS)
valid_acc_values = np.zeros(N_EPOCHS)

In [ ]:
# best_valid_acc = 0

# train_loss_values = np.zeros(N_EPOCHS)
# train_acc_values = np.zeros(N_EPOCHS)

# valid_loss_values = np.zeros(N_EPOCHS)
# valid_acc_values = np.zeros(N_EPOCHS)

for epoch in range(N_EPOCHS):

    train_loss, train_acc = train(CNN_model, train_loader, optimizer, criterion)

    train_loss_values[N_EPOCHS - 1] = train_loss
    train_acc_values[N_EPOCHS - 1] = train_acc

    torch.save(CNN_model.state_dict(), MODEL_PATH)
    torch.save(optimizer.state_dict(), MODEL_PATH + '.optim')
    if True: # It was epoch % 2 == 0:
        print (f'| Epoch: {epoch+1:02} | Train Loss: {train_loss:.3f} | Train Acc: {train_acc*100:.2f}% |')
        file = open("./losses.txt", "a")
        file.write("epoch: " + str(epoch) + " train_loss: " + str(train_loss) + '\n')

    if True: # It was epoch % 5 == 0:
        print ("Begin validation")
        valid_loss, valid_acc = evaluate(CNN_model, dev_loader, criterion)

        valid_loss_values[N_EPOCHS - 1] = valid_loss
        valid_acc_values[N_EPOCHS - 1] = valid_acc

        print (f'| Val. Loss: {valid_loss:.3f} | Val. Acc: {valid_acc*100:.2f}% |')

        if valid_acc > best_valid_acc:
            torch.save(CNN_model.state_dict(), MODEL_PATH)
            torch.save(optimizer.state_dict(), MODEL_PATH + '.optim')
            best_valid_acc = valid_acc

        file = open("./losses.txt", "a")
        file.write("epoch: " + str(epoch) + " train_loss: " + str(train_loss))

| Epoch: 01 | Train Loss: 0.668 | Train Acc: 65.52% |
Begin validation
| Val. Loss: 2.078 | Val. Acc: 66.39% |
| Epoch: 02 | Train Loss: 0.597 | Train Acc: 70.99% |
Begin validation
| Val. Loss: 904.895 | Val. Acc: 74.34% |
| Epoch: 03 | Train Loss: 0.573 | Train Acc: 72.62% |
Begin validation
| Val. Loss: 15.983 | Val. Acc: 67.20% |
| Epoch: 04 | Train Loss: 0.563 | Train Acc: 73.47% |
Begin validation
| Val. Loss: 0.601 | Val. Acc: 76.25% |
| Epoch: 05 | Train Loss: 0.552 | Train Acc: 74.15% |
Begin validation
| Val. Loss: 57216.306 | Val. Acc: 76.46% |
| Epoch: 06 | Train Loss: 0.543 | Train Acc: 74.73% |
Begin validation
| Val. Loss: 14505.082 | Val. Acc: 74.17% |
| Epoch: 07 | Train Loss: 0.537 | Train Acc: 75.10% |
Begin validation
| Val. Loss: 181.478 | Val. Acc: 77.20% |
| Epoch: 08 | Train Loss: 0.530 | Train Acc: 75.62% |
Begin validation
| Val. Loss: 168.771 | Val. Acc: 78.43% |
| Epoch: 09 | Train Loss: 0.523 | Train Acc: 75.99% |
Begin validation
| Val. Loss: 255.815 | Val

https://pytorch.org/tutorials/beginner/saving_loading_models.html#saving-loading-a-general-checkpoint-for-inference-and-or-resuming-training

In [ ]:
# ##Ehsan Added 14010624##
# ## Begin
# # CNN_model = TheModelClass(*args, **kwargs)
# # optimizer = TheOptimizerClass(*args, **kwargs)

# checkpoint = torch.load(MODEL_PATH)
# CNN_model.load_state_dict(checkpoint['model_state_dict'])
# optimizer.load_state_dict(checkpoint['optimizer_state_dict'])
# epoch = checkpoint['epoch']
# loss = checkpoint['loss']

# model.eval()
# # - or -
# # model.train()
# ## End


# # CNN_model.load_state_dict(torch.load(MODEL_PATH))
# # CNN_model.load_state_dict(torch.load(MODEL_PATH + '.optim'))
# # CNN_model.eval()


In [ ]:
################## Save to LOAD again ##################
for epoch in range(N_EPOCHS):

    train_loss, train_acc = train(CNN_model, train_loader, optimizer, criterion)

    train_loss_values[epoch] = train_loss
    train_acc_values[epoch] = train_acc

#    torch.save(CNN_model.state_dict(), MODEL_PATH)
#    torch.save(optimizer.state_dict(), MODEL_PATH + '.optim')


    torch.save({
                'epoch': epoch,
                'model_state_dict': CNN_model.state_dict(),
                'optimizer_state_dict': optimizer.state_dict(),
                'loss': loss,
                ...
                }, MODEL_PATH)


    if True: # It was epoch % 2 == 0:
        print (f'| Epoch: {epoch+1:02} | Train Loss: {train_loss:.3f} | Train Acc: {train_acc*100:.2f}% |')
        file = open("./losses.txt", "a")
        file.write("epoch: " + str(epoch) + " train_loss: " + str(train_loss) + '\n')

    if True: # It was epoch % 5 == 0:
        # print ("Begin validation")
        valid_loss, valid_acc = evaluate(CNN_model, dev_loader, criterion)

        valid_loss_values[epoch] = valid_loss
        valid_acc_values[epoch] = valid_acc

        print (f'|          Val. Loss: {valid_loss:.3f} | Val. Acc: {valid_acc*100:.2f}% |')
        print ('-----------------------------------------------------------------')

        if valid_acc > best_valid_acc:
            # torch.save(CNN_model.state_dict(), MODEL_PATH)
            # torch.save(optimizer.state_dict(), MODEL_PATH + '.optim')

            torch.save({
                        'epoch': epoch,
                        'model_state_dict': CNN_model.state_dict(),
                        'optimizer_state_dict': optimizer.state_dict(),
                        'loss': loss,
                        ...
                        }, MODEL_PATH)


            best_valid_acc = valid_acc

        file = open("./losses.txt", "a")
        file.write("epoch: " + str(epoch) + " train_loss: " + str(train_loss))

In [ ]:
# CNN_model = TheModelClass(*args, **kwargs)

| Epoch: 01 | Train Loss: 0.591 | Train Acc: 71.40% |
|          Val. Loss: 9.735 | Val. Acc: 73.38% |
-----------------------------------------------------------------
| Epoch: 01 | Train Loss: 0.546 | Train Acc: 74.53% |
|          Val. Loss: 609.560 | Val. Acc: 78.98% |
-----------------------------------------------------------------
| Epoch: 01 | Train Loss: 0.539 | Train Acc: 74.92% |
|          Val. Loss: 496.271 | Val. Acc: 77.27% |
-----------------------------------------------------------------
| Epoch: 01 | Train Loss: 0.531 | Train Acc: 75.34% |
|          Val. Loss: 35.134 | Val. Acc: 80.23% |
-----------------------------------------------------------------


In [ ]:
len(train_loss_values)

In [ ]:
import matplotlib.pyplot as plt
plt.style.use('ggplot')

In [ ]:
print('train_acc_values:',train_acc_values,'\n')
print('valid_acc_values:',valid_acc_values,'\n')

In [ ]:
x = range(0, (N_EPOCHS))

# train_loss_values
# train_acc_values
# valid_loss_values
# valid_acc_values
# xpoints = np.array(train_loss_values, train_loss_values)

plt.figure(figsize=(12,5), dpi=80)
plt.subplot(1, 2, 1)
plt.plot(x, train_acc_values, 'b', label='Traning acc')
plt.plot(x, valid_acc_values, 'r', label='Validation acc')
plt.title('Traning and Validation accuracy')
plt.legend()

plt.subplot(1, 2, 2)
plt.plot(x, train_loss_values, 'b', label='Traning loss')
plt.plot(x, valid_loss_values, 'r', label='Validation loss')
plt.title('Traning and Validation loss')
plt.legend()




In [ ]:

plt.axis([0.7, 1, 0.6, 0.8])
plt.xlabel('Training Accuracy')
plt.ylabel('Validation Accuracy')
plt.plot(train_acc_values, valid_acc_values)
plt.show()